packages/contracts-bedrock/src/Safe/LivenessModule.sol

```diff
@@ -0,0 +1,216 @@
+
+    /// @notice The minimum number of owners before ownership of the safe is transferred to the fallback owner.
+    ///         This can be updated by replacing with a new module.
+    uint256 internal immutable MIN_OWNERS;
+
+    /// @notice The fallback owner of the Safe
+    ///         This can be updated by replacing with a new module.
+    address internal immutable FALLBACK_OWNER;
+
+    /// @notice The storage slot used in the safe to store the guard address
+    ///         keccak256("guard_manager.guard.address")
+    uint256 internal constant GUARD_STORAGE_SLOT = 0x4a204f620c8c5ccdca3fd54d003badd85ba500436a431f0cbda4f558c93c34c8;
+
+    /// @notice Semantic version.
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    // Constructor to initialize the Safe and baseModule instances
+    constructor(
+        Safe _safe,
+        LivenessGuard _livenessGuard,
+        uint256 _livenessInterval,
+        uint256 _minOwners,
+        address _fallbackOwner
+    ) {
+        SAFE = _safe;
+        LIVENESS_GUARD = _livenessGuard;
+        LIVENESS_INTERVAL = _livenessInterval;
+        FALLBACK_OWNER = _fallbackOwner;
+        MIN_OWNERS = _minOwners;
+        address[] memory owners = _safe.getOwners();
+        require(_minOwners < owners.length, "LivenessModule: minOwners must be less than the number of owners");
+        require(
+            _safe.getThreshold() == get75PercentThreshold(owners.length),
+            "LivenessModule: Safe must have a threshold of 75% of the number of owners"
+        );
+    }
+
+    /// @notice This function can be called by anyone to remove a set of owners that have not signed a transaction
+    ///         during the liveness interval. If the number of owners drops below the minimum, then all owners
+    ///         must be removed.
+    /// @param _previousOwners The previous owners in the linked list of owners
+    /// @param _ownersToRemove The owners to remove
+    function removeOwners(address[] memory _previousOwners, address[] memory _ownersToRemove) external {
+        require(_previousOwners.length == _ownersToRemove.length, "LivenessModule: arrays must be the same length");
+
+        // We will remove at least one owner, so we'll initialize the newOwners count to the current number of owners
+        uint256 ownersCount = SAFE.getOwners().length;
+        for (uint256 i = 0; i < _previousOwners.length; i++) {
+            ownersCount--;
+            if (ownersCount >= MIN_OWNERS) {
+                require(_canRemove(_ownersToRemove[i]), "LivenessModule: the owner to remove has signed recently");
+            }
+
+            _removeOwner({
+                _prevOwner: _previousOwners[i],
+                _ownerToRemove: _ownersToRemove[i],
+                _newOwnersCount: ownersCount,
+                _newThreshold: get75PercentThreshold(ownersCount)
+            });
+        }
+        _verifyFinalState();
```

Comment:
We have to be sure that a malicious input cannot cause issues here. An example would be the wrong ordered inputs

Reply:
I think I see your point.

But otoh, properly ordering the inputs is up to the user, if any invalid condition results, then either the safe or verifyFinalState should revert. There are tests for both. 

Definitely if this module is not asserting on some invariant it needs to be added, but it should leave it up to the Safe to validate the ordering of owner accounts in its calls. 

Am I following or is there something else you're getting at?

Reply:
Resolving. A few notes for posterity: 

1. Reverts in this code path of the module are OK, it's the Guard where that would be very dangerous.
2. A [very thorough test](https://github.com/ethereum-optimism/optimism/blob/jm/safe-liveness/packages/contracts-bedrock/test/LivenessModule.t.sol#L391) of the frei-pi checks has been written